Regression: package was not installed in a way that PackageLoader understands.

## Expected Behavior
PackageLoader should continue to load from a namespace package even under pytest.

## Actual Behavior
Starting with jinja2 2.11, the tests for pmxbot have [started failing](https://travis-ci.org/yougov/pmxbot/jobs/647234617?utm_medium=notification&utm_source=github_status). They fail when pytest attempts to rewrite the modules for running under test and in doing so, cause the `jinja2.loaders.PackageLoader` not to recognize the 'pmxbot.web' package. Note that `pmxbot` is a setuptools namespace package.

## Full Traceback
```pytb
pmxbot/web/__main__.py:1: in <module>
    import pmxbot.web.viewer
<frozen importlib._bootstrap>:991: in _find_and_load
    ???
<frozen importlib._bootstrap>:975: in _find_and_load_unlocked
    ???
<frozen importlib._bootstrap>:671: in _load_unlocked
    ???
.tox/python/lib/python3.8/site-packages/_pytest/assertion/rewrite.py:143: in exec_module
    exec(co, module.__dict__)
pmxbot/web/viewer.py:24: in <module>
    jenv = jinja2.Environment(loader=jinja2.loaders.PackageLoader('pmxbot.web'))
.tox/python/lib/python3.8/site-packages/jinja2/loaders.py:288: in __init__
    raise ValueError(
E   ValueError: The 'pmxbot.web' package was not installed in a way that PackageLoader understands.
```

The error is not triggered when simply importing the package naturally; it only seems to happen during the pytest rewrite.

## Your Environment
* Python version: 3.8.1
* Jinja version: 2.11.x

Downgrading to Jinja 2.10.3 works around the failure. I suspect #970 is implicated.

--- a/CHANGES.rst
+++ b/CHANGES.rst
@@ -9,6 +9,8 @@ Unreleased
 -   Bump MarkupSafe dependency to >=1.1.
 -   Bump Babel optional dependency to >=2.1.
 -   Remove code that was marked deprecated.
+-   Use :pep:`451` API to load templates with
+    :class:`~loaders.PackageLoader`. :issue:`1168`
 
 
 2.11.2

--- a/src/jinja2/loaders.py
+++ b/src/jinja2/loaders.py
@@ -1,10 +1,11 @@
 """API and implementations for loading templates from different data
 sources.
 """
+import importlib.util
 import os
-import pkgutil
 import sys
 import weakref
+import zipimport
 from collections import abc
 from hashlib import sha1
 from importlib import import_module
@@ -253,21 +254,19 @@ class PackageLoader(BaseLoader):
         # Make sure the package exists. This also makes namespace
         # packages work, otherwise get_loader returns None.
         import_module(package_name)
-        self._loader = loader = pkgutil.get_loader(package_name)
+        spec = importlib.util.find_spec(package_name)
+        self._loader = spec.loader
 
-        # Zip loader's archive attribute points at the zip.
-        self._archive = getattr(loader, "archive", None)
+        self._archive = None
         self._template_root = None
-
-        if hasattr(loader, "get_filename"):
-            # A standard directory package, or a zip package.
-            self._template_root = os.path.join(
-                os.path.dirname(loader.get_filename(package_name)), package_path
-            )
-        elif hasattr(loader, "_path"):
-            # A namespace package, limited support. Find the first
-            # contributor with the template directory.
-            for root in loader._path:
+        if isinstance(spec.loader, zipimport.zipimporter):
+            self._archive = spec.loader.archive
+            pkgdir = next(iter(spec.submodule_search_locations))
+            self._template_root = os.path.join(pkgdir, package_path)
+        elif spec.submodule_search_locations:
+            # this will be one element for "packages" and multiple for
+            # namespace packages
+            for root in spec.submodule_search_locations:
                 root = os.path.join(root, package_path)
 
                 if os.path.isdir(root):

--- a/src/jinja2/loaders.py
+++ b/src/jinja2/loaders.py
@@ -255,17 +255,17 @@ class PackageLoader(BaseLoader):
         # packages work, otherwise get_loader returns None.
         import_module(package_name)
         spec = importlib.util.find_spec(package_name)
-        self._loader = spec.loader
-
+        self._loader = loader = spec.loader
         self._archive = None
         self._template_root = None
-        if isinstance(spec.loader, zipimport.zipimporter):
-            self._archive = spec.loader.archive
+
+        if isinstance(loader, zipimport.zipimporter):
+            self._archive = loader.archive
             pkgdir = next(iter(spec.submodule_search_locations))
             self._template_root = os.path.join(pkgdir, package_path)
         elif spec.submodule_search_locations:
-            # this will be one element for "packages" and multiple for
-            # namespace packages
+            # This will be one element for regular packages and multiple
+            # for namespace packages.
             for root in spec.submodule_search_locations:
                 root = os.path.join(root, package_path)
 

--- a/tests/test_loader.py
+++ b/tests/test_loader.py
@@ -352,21 +352,13 @@ def test_package_zip_list(package_zip_loader):
     assert package_zip_loader.list_templates() == ["foo/test.html", "test.html"]
 
 
-def test_pep_451_import_hook(tmp_path):
-    package_name = "_my_pep451_pkg"
-    pkg = tmp_path.joinpath(package_name)
-    pkg.mkdir()
-    pkg.joinpath("__init__.py").touch()
-    templates = pkg.joinpath("templates")
-    templates.mkdir()
-    templates.joinpath("foo.html").write_text("hello world")
-
+def test_pep_451_import_hook():
     class ImportHook(importlib.abc.MetaPathFinder, importlib.abc.Loader):
         def find_spec(self, name, path=None, target=None):
-            if name != package_name:
+            if name != "res":
                 return None
-            path = [str(tmp_path)]
-            spec = importlib.machinery.PathFinder.find_spec(name, path=path)
+
+            spec = importlib.machinery.PathFinder.find_spec(name)
             return importlib.util.spec_from_file_location(
                 name,
                 spec.origin,
@@ -382,9 +374,10 @@ def test_pep_451_import_hook(tmp_path):
 
     # ensure we restore `sys.meta_path` after putting in our loader
     before = sys.meta_path[:]
+
     try:
         sys.meta_path.insert(0, ImportHook())
-        package_loader = PackageLoader(package_name)
-        assert package_loader.list_templates() == ["foo.html"]
+        package_loader = PackageLoader("res")
+        assert "test.html" in package_loader.list_templates()
     finally:
         sys.meta_path[:] = before

--- a/tests/test_loader.py
+++ b/tests/test_loader.py
@@ -1,3 +1,6 @@
+import importlib.abc
+import importlib.machinery
+import importlib.util
 import os
 import platform
 import shutil
@@ -347,3 +350,41 @@ def test_package_zip_source(package_zip_loader, template, expect):
 )
 def test_package_zip_list(package_zip_loader):
     assert package_zip_loader.list_templates() == ["foo/test.html", "test.html"]
+
+
+def test_pep_451_import_hook(tmp_path):
+    package_name = "_my_pep451_pkg"
+    pkg = tmp_path.joinpath(package_name)
+    pkg.mkdir()
+    pkg.joinpath("__init__.py").touch()
+    templates = pkg.joinpath("templates")
+    templates.mkdir()
+    templates.joinpath("foo.html").write_text("hello world")
+
+    class ImportHook(importlib.abc.MetaPathFinder, importlib.abc.Loader):
+        def find_spec(self, name, path=None, target=None):
+            if name != package_name:
+                return None
+            path = [str(tmp_path)]
+            spec = importlib.machinery.PathFinder.find_spec(name, path=path)
+            return importlib.util.spec_from_file_location(
+                name,
+                spec.origin,
+                loader=self,
+                submodule_search_locations=spec.submodule_search_locations,
+            )
+
+        def create_module(self, spec):
+            return None  # default behaviour is fine
+
+        def exec_module(self, module):
+            return None  # we need this to satisfy the interface, it's wrong
+
+    # ensure we restore `sys.meta_path` after putting in our loader
+    before = sys.meta_path[:]
+    try:
+        sys.meta_path.insert(0, ImportHook())
+        package_loader = PackageLoader(package_name)
+        assert package_loader.list_templates() == ["foo.html"]
+    finally:
+        sys.meta_path[:] = before

